server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -466,32 +472,111 @@ public synchronized void verifyIndexMetadata(IndexMetadata metadata, IndexMetada
         }
     }
 
+
+    private CompletableFuture<IndexShard> createShard(ClusterState originalState,
+                                                      IndexService indexService,
+                                                      ShardRouting shardRouting,
+                                                      Consumer<ShardId> globalCheckpointSyncer,
+                                                      RetentionLeaseSyncer retentionLeaseSyncer,
+                                                      @Nullable Iterator<TimeValue> backoff) {
+        try {
+            IndexShard indexShard = indexService.createShard(shardRouting, globalCheckpointSyncer, retentionLeaseSyncer);
+            return CompletableFuture.completedFuture(indexShard);
+        } catch (ShardLockObtainFailedException e) {
+            CompletableFuture<IndexShard> result = new CompletableFuture<>();
+            // ~1.1hours in total; Retries take never more than 5000ms
+            int firstDelayInMS = 50;
+            int maxRetries = 1000;
+            int maxDelayInMS = 5000;
+            Iterator<TimeValue> backoffIt = backoff == null
+                ? new LimitedExponentialBackoff(firstDelayInMS, maxRetries, maxDelayInMS).iterator()
+                : backoff;
+            TimeValue delay = backoffIt.next();
+            if (LOGGER.isWarnEnabled() && delay.millis() == maxDelayInMS) {
+                LOGGER.warn("Repeated attempts to acquire shardLock for {}. Retrying again in {}", shardRouting.shardId(), delay);
+            } else if (LOGGER.isDebugEnabled() && delay.millis() > 150) {
+                LOGGER.debug("Repeated attempts to acquire shardLock for {}. Retrying again in {}", shardRouting.shardId(), delay);
+            }
+            threadPool.scheduleUnlessShuttingDown(delay, ThreadPool.Names.SAME, () -> {
```

Comment:
Is it by intend to block all cluster state update tasks on retries? (Using same threadpool, which is uses the single thread pool one of the `ClusterApplierService`  https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java#L337)
I'm not completely sure on this, sounds somehow valid....

Reply:
How is it blocking?
Using `SAME` for the schedule call means the scheduler won't spawn the command in a separate executor. It doesn't block the scheduler itself.
And the command has a manual delegate to the ClusterApplierService executor, via `runOnApplierThread`

https://github.com/crate/crate/blob/99d183032d4b5763a24a84ddf909806d94df4081/server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java#L84